Feat: Add CBR DA

dishankkalra23 opened the pull request:
### Description

<!--- Replace this text with a summary of the changes in this PR. Include why the changes are needed and context about the changes. List required dependencies. If there is a Git issue for the change, please link to it. --->

This PR adds a basic DA for CBR, creating CBR zones and associated rules with context

### Release required?
<!--- Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning). --->

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

<!--- If a release is required, replace this text with information that users need to know about the release. Write the release notes to help users understand the changes, and include information about how to update from the previous version.

Your notes help the merger write the commit message for the PR that is published in the release notes for the module. --->

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

dishankkalra23 commented:
/run pipeline

dishankkalra23 commented:
/run pipeline

dishankkalra23 commented:
/run pipeline

dishankkalra23 commented:
/run pipeline

Ak-sky commented on reference-architecture/cbr.svg:
Please take reference from this diagram- https://cloud.ibm.com/docs/account?topic=account-context-restrictions-whatis.
It should should show the flow of network from CBR zone to the target service. Also the other resources which are getting created by the DA.

Ak-sky reviewed (changes requested):
- In my opinion, DA should call the FSCloud module as the module has more flexibility to create CBR rules and zones, and disable the pre-wired rules if not needed by default.
- Add test for the DA.

dishankkalra23 commented:
`In my opinion, DA should call the FSCloud module as the module has more flexibility to create CBR rules and zones, and disable the pre-wired rules if not needed by default.`

@Ak-sky Thanks for the suggestion. However, introducing the FSCloud module here would add several additional variables and configuration options, which we have observed many service teams may not require especially those preferring to manage their own custom CBR rules.

To keep this DA simple and easier to consume, it might be better to keep it as a `basic` solution and introducing another DA for FSCloud module.

dishankkalra23 commented:
Added tests: `TestBasicDAInSchematics` and `TestRunUpgradeBasicDAInSchematics`

- In `TestBasicDAInSchematics` I have included a dummy non-routable private IP `192.168.99.99` to create a zone so it won’t cause issues during testing. Please update it with an appropriate IP if needed, as long as it doesn't disrupt the GE account.

- For `cbr_rules`, I could not add a dummy value because the `resources` attribute expects a real Cloud resource where restrictions needs to be applied. Since I wasn’t sure which permanent service from the GE account can be safely used for this I have set the `cbr_rules` variable to an empty map by default.

dishankkalra23 commented:
- The `TestBasicDAInSchematics` test passed successfully.

```
MacBook-Pro:terraform-ibm-cbr dishankkalra$ make run-tests-local RUN=TestBasicDAInSchematics
cd tests && go test -run TestBasicDAInSchematics -count=1 -v -timeout 600m
=== RUN   TestBasicDAInSchematics
=== PAUSE TestBasicDAInSchematics
=== CONT  TestBasicDAInSchematics
2025/06/26 23:35:35 Region au-syd VPC count: 0
2025/06/26 23:35:35 --- new best region is au-syd
2025/06/26 23:35:35 Best region was found!: au-syd
    tests.go:330: [SCHEMATICS] Random Workspace region chosen: eu
    schematics.go:136: [SCHEMATICS] Schematics API for region eu: https://eu.schematics.cloud.ibm.com
    tests.go:100: [SCHEMATICS] Creating Test Workspace
    tests.go:106: [SCHEMATICS] Workspace Created: cbr-da-iss3yg (eu-gb.workspace.cbr-da-iss3yg.453605e2)
    schematics.go:233: [SCHEMATICS] Creating TAR file
    schematics.go:239: [SCHEMATICS] Uploading TAR file
    schematics.go:557: [SCHEMATICS] ... still waiting for job TAR_WORKSPACE_UPLOAD to complete: 1 minutes
    schematics.go:566: [SCHEMATICS] The status of job TAR_WORKSPACE_UPLOAD is: COMPLETED
    tests.go:136: [SCHEMATICS] Updating Workspace Variablestore
    tests.go:152: [SCHEMATICS] Starting PLAN job ...
    schematics.go:557: [SCHEMATICS] ... still waiting for job PLAN to complete: 1 minutes
    schematics.go:566: [SCHEMATICS] The status of job PLAN is: COMPLETED
    tests.go:186: [SCHEMATICS] Starting APPLY job ...
    schematics.go:557: [SCHEMATICS] ... still waiting for job APPLY to complete: 1 minutes
    schematics.go:566: [SCHEMATICS] The status of job APPLY is: COMPLETED
    tests.go:245: [SCHEMATICS] Starting CONSISTENCY PLAN job ...
    schematics.go:557: [SCHEMATICS] ... still waiting for job PLAN to complete: 1 minutes
    schematics.go:566: [SCHEMATICS] The status of job PLAN is: COMPLETED
    tests.go:438: [SCHEMATICS] Starting DESTROY job ...
    schematics.go:557: [SCHEMATICS] ... still waiting for job DESTROY to complete: 1 minutes
    schematics.go:566: [SCHEMATICS] The status of job DESTROY is: COMPLETED
    tests.go:465: [SCHEMATICS] Deleting Workspace
--- PASS: TestBasicDAInSchematics (345.95s)
PASS
ok      github.com/terraform-ibm-modules/terraform-ibm-cbr      347.255s
```

- The `TestRunUpgradeBasicDAInSchematics` test is currently failing, even after multiple retries. Based on the logs, the error appears from schematics during the TAR upload step.

```
=== RUN   TestRunUpgradeBasicDAInSchematics

schematics.go:566: The status of job TAR_WORKSPACE_UPLOAD is: FAILED

Error: Expected nil, but got: error setting workspace tar file: tar upload has failed with status FAILED
--- FAIL: TestRunUpgradeBasicDAInSchematics (87.95s)
FAIL
exit status 1
```

Ak-sky commented:
> Added tests: `TestBasicDAInSchematics` and `TestRunUpgradeBasicDAInSchematics`
> 
> * In `TestBasicDAInSchematics` I have included a dummy non-routable private IP `192.168.99.99` to create a zone so it won’t cause issues during testing. Please update it with an appropriate IP if needed, as long as it doesn't disrupt the GE account.
> * For `cbr_rules`, I could not add a dummy value because the `resources` attribute expects a real Cloud resource where restrictions needs to be applied. Since I wasn’t sure which permanent service from the GE account can be safely used for this I have set the `cbr_rules` variable to an empty map by default.

Zone is fine, also it is ok to create rules for any target service as long as `enforcement_mode` is report or disabled.
So you can add sample rules as well in the PR test.

dishankkalra23 commented:
1. Added the mocked data for `cbr_rules` and the rules are getting created and destroyed successfully during test. Included sample rules for both Postgres and Redis services with `enforcement_mode` set to `disabled`. These will be created in `ge_dev_account`, where PR tests are executed within GE.

2. Additionally, I have updated the variable definition to set enforcement_mode to disabled by default if not provided by user.

dishankkalra23 commented:
1. `TestBasicDAInSchematics` passed

```
=== NAME  TestBasicDAInSchematics
    schematics.go:557: [SCHEMATICS] ... still waiting for job APPLY to complete: 2 minutes
    schematics.go:566: [SCHEMATICS] The status of job APPLY is: COMPLETED
    tests.go:245: [SCHEMATICS] Starting CONSISTENCY PLAN job ...

=== NAME  TestBasicDAInSchematics
    schematics.go:557: [SCHEMATICS] ... still waiting for job DESTROY to complete: 2 minutes
    schematics.go:566: [SCHEMATICS] The status of job DESTROY is: COMPLETED
    tests.go:465: [SCHEMATICS] Deleting Workspace
--- PASS: TestBasicDAInSchematics (609.86s)
```

2. `TestRunUpgradeBasicDAInSchematics` is failing with same error as mentioned above in other comment: https://github.com/terraform-ibm-modules/terraform-ibm-cbr/pull/674#issuecomment-3009439883

```
=== NAME  TestRunUpgradeBasicDAInSchematics
    tests.go:488: [SCHEMATICS] Removing temp directory for upgrade test: /tmp/terraform-base-cbr-da-k6m85h914571408
    pr_test.go:603: 
        	Error Trace:	/__w/terraform-ibm-cbr/terraform-ibm-cbr/tests/pr_test.go:603
        	Error:      	Expected nil, but got: &fmt.wrapError{msg:"error setting workspace tar file: tar upload has failed with status FAILED - [ cbr-da-k6m85h (us-east.workspace.cbr-da-k6m85h.30819cd1) ]", err:(*errors.errorString)(0xc000d92c90)}
        	Test:       	TestRunUpgradeBasicDAInSchematics
        	Messages:   	This should not have errored
--- FAIL: TestRunUpgradeBasicDAInSchematics (220.84s)
```

Ak-sky commented:
/run pipeline

Ak-sky commented:
<img width="1699" alt="image" src="https://github.com/user-attachments/assets/88c59f15-21db-46cd-ac7c-390fc687668a" />

ocofaigh reviewed (changes requested):
Left a few comments. I'm also not sure on the variation name "Basic" - is it really a basic variation? Seems advanced to me. Should it be called "Fully configurable"

Ak-sky commented:
> Left a few comments. I'm also not sure on the variation name "Basic" - is it really a basic variation? Seems advanced to me. Should it be called "Fully configurable"

We can name it to fully configurable but as it is not consuming fscloud module which has more flexibility hence named this variations as basic, same was discussed in the last playback call.

maheshwarishikha reviewed (changes requested):
Also,
- catalog_onboard pipeline file is missing
- release.rc is not updated

Ak-sky commented:
/run pipeline

maheshwarishikha commented:
/run pipeline

terraform-ibm-modules-ops commented:
:tada: This PR is included in version 1.32.0 :tada:

The release is available on:
- [GitHub release](https://github.com/terraform-ibm-modules/terraform-ibm-cbr/releases/tag/v1.32.0)
- `v1.32.0`

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
